Better Package Style layout (#71)
pyansys/dev-guide: doc/source/packaging/build-systems.rst

.. _ref_build_system:

############
Build System
############

> The build system is a fundamental tool when it comes to packaging Python

Reviewer: ```suggestion
The build system is a fundamental tool for packaging Python
```

> libraries. It allows to generate distribution files which can be shared with

Reviewer: ```suggestion
libraries. It generates distribution files that can be shared with
```

> The build system is a fundamental tool when it comes to packaging Python
> libraries. It allows to generate distribution files which can be shared with
> users and developers. 

Reviewer: ```suggestion
The build system is a fundamental tool for packaging Python
libraries. It generates distribution files that can be shared with
users and developers. 
```

> Artifacts
> ================

Reviewer 2: ```suggestion
Artifacts
=========
```

Author: Good catch!

> The Build System

Reviewer: @jorgepiloto I'm a bit  confused as to what seems an interchangeable use of build system versus build system. Is there a way that this can be consistent or clarified?

Author: Thanks for pointing this out @jorgepiloto. You are right, having "build system" in the header of the page and in here is confusing.

We could rename this section as "Artifacts", as it is a short title and these are the end goal of the build system.

Reviewer: My comment has been edited. I wanted to point out that there were instances of "build system" and "build-system" and I was hoping they could be made consistent.

> The Build System

Author: ```suggestion
Artifacts
```

> The build-system allows maintainers to generate artifacts for their Python

Reviewer: ```suggestion
The build system allows maintainers to generate artifacts for their Python
```

> libraries. The word artifacts refers to both wheel and source files:

Reviewer: ```suggestion
libraries. Here, `artifacts` refers to both wheel and source files:
```

- ``Wheel files`` have the ``*.whl`` file extension.
- ``Source files`` have the ``*.tar.gz`` or ``*.zip`` extension.

> These are the files to be uploaded to `PyPI`_ when releasing a new version of a

Reviewer: ```suggestion
These are the files to upload to `PyPI`_ when releasing a new version of a
```

PyAnsys project.

.. warning::

   Not all files are included by default in the source distribution. A
   `MANIFEST.in`_ is required at the root of the project to specify additional
   files.

> The interaction between the maintainer and the build-system is performed using a

Reviewer: ```suggestion
The interaction between the maintainer and the build system is performed using a
```

build-system tool. This tool provides a frontend and a backend. The maintainers

> trigger the frontend which then calls the backend. Then the backend reads the

Reviewer: ```suggestion
trigger the frontend, which then calls the backend. The backend then reads the
```

> build-system tool. This tool provides a frontend and a backend. The maintainers
> trigger the frontend, which then calls the backend. The backend then reads the
> project directory and generates the artifacts, as :numref:`build system diag`:

Reviewer: ```suggestion
build system tool. This tool provides both a frontend and a backend. The maintainers
trigger the frontend, which then calls the backend to read the
project directory and generate the artifacts, as :numref:`build system diag`:
```

Reviewer: @jorgepiloto Should the ", as" at the end of the last sentence be ", as shown in this diagram." I'm not sure how this directive as is renders.

Author: Thanks for spotting this, @PipKat. The ":numref:`name of reference`" render as `Fig.X`, where `X` is the ith-figure of the docs. Therefore, this could be refactored as:

```suggestion
build-system tool. This tool provides a frontend and a backend. The maintainers
trigger the frontend, which then calls the backend. The backend then reads the
project directory and generates the artifacts, as shown in :numref:`build system diag`:
```

.. include:: diagrams/build_system_diagram.rst

> The PEP 517 and PEP 518

Reviewer: ```suggestion
PEP 517 and PEP 518
```

> The PEP 517 and PEP 518
> =======================

Reviewer: ```suggestion
PEP 517 and PEP 518
===================
```

For a long time, the ``setup.py`` file was the only way of specifying the

> project structure, metadata and installation workflow to be followed by `pip`_.

Reviewer: ```suggestion
project structure, metadata, and installation workflow that `pip`_ was to follow.
```

However, having to execute a Python file when installing a Python package
introduced the following problems:

- It was not possible to know which dependencies required the ``setup.py`` file
  to be properly executed.

- The default Python package installer, `pip`_, expected `setuptools`_ to be the
  default build system tool, excluding others like `flit`_ and `poetry`_.

> Previous reasons lead to the acceptance of the `PEP 517`_ and `PEP 518`_.

Reviewer: ```suggestion
These problems led to the acceptance of `PEP 517`_ and `PEP 518`_.
```

PEP 517
-------

> `PEP 517` allows Python developers to specify the build-backend tool for

Reviewer 2: ```suggestion
`PEP 517`_ allows Python developers to specify the build-backend tool for
```

> `PEP 517`_ allows Python developers to specify the build-backend tool for
> generating artifacts. As seen in the :numref:`build system diag` figure, these
> are the most popular backends:

Reviewer: ```suggestion
`PEP 517`_ allows Python developers to specify the build backend tool for
generating artifacts. The previous :numref:`build system diag` diagram shows the
most popular backends:
```

> The `PEP 517` allowed Python developers to specify the build-backend tool to
> generate the artifacts. As seen in :numref:`build system diagram` figure, the
> most popular backends are provided by:

Reviewer: ```suggestion
`PEP 517` allows Python developers to specify the build-backend tool for
generating artifacts. As seen in the :numref:`build system diagram` figure, these
are the most popular backends:
```

> - `Setuptools`_ , while very popular, lacks build time dependency declaration

Reviewer: ```suggestion
- `Setuptools`_ , while very popular, lacks build-time dependency declaration
```

> - `Setuptools`_ , while very popular, lacks build time dependency declaration
>   and is difficult to extend.

Reviewer: ```suggestion
- `Setuptools`_ , while very popular, lacks the ability to declare build time dependencies
  and is difficult to extend.
```

Reviewer: @jorgepiloto If "build time dependency declaration" is a common developer term, you can ignore this comment. The general guideline is to limit the adjectives that modify a noun to two to keep it more readable, comprehensible, and, translatable. However, exceptions exist for terms that are commonly used and understood.

> - `Setuptools`_ very popular but lacks from build time dependency declaration
>   and it is difficult to extend.

Reviewer: ```suggestion
- `Setuptools`_ , while very popular, lacks build time dependency declaration
  and is difficult to extend.
```

- `Flit`_ is a lightweight build system tool for Python.
- `Poetry`_ focuses on dependency management and environment isolation.

> The `PEP 517` introduced the ``build-backend`` key inside the

Reviewer: ```suggestion
`PEP 517` introduced the ``build-backend`` key inside the
```

``[build-system]`` table in the ``pyproject.toml``.

PEP 518
-------

> In addition to the ``setup.py`` file, a new project file named
> ``pyproject.toml`` file was specified by the `PEP 518`_. The main goal of this
> file was to specify build time dependencies. However, some build system tools
> like `flit`_ or `poetry`_ are able to specify all the project metadata inside
> the ``pyproject.toml`` file and eliminate the usage of the ``setup.py``.

Reviewer: ```suggestion
In addition to the ``setup.py`` file, `PEP 518`_ specifies a project file named
``pyproject.toml``. The main goal of this file is to specify build-time dependencies.
However, some build system tools like `flit`_ or `poetry`_ are able to specify all
project metadata inside the ``pyproject.toml`` file and eliminate usage of the 
``setup.py`` file.
```

> To specify the build time requirements, the ``[build-system]`` table needs to be

Reviewer: ```suggestion
To specify the build time requirements, the ``[build-system]`` table must be
```

declared in the ``pyproject.toml`` file. Within it, the ``requires`` key is

> assigned to a list of strings, each one of those being the build time
> requirements.

Reviewer: ```suggestion
assigned to a list of strings, which are the build time
requirements.
```

> The combination of `PEP 517`_ and `PEP 518`_ leads to the following syntax in te
> ``pyproject.toml`` files:

Reviewer: ```suggestion
The combination of `PEP 517`_ and `PEP 518`_ leads to the following syntax in a
``pyproject.toml`` file:
```

.. code:: toml

   [build-system]
   requires = ["flit"] # Defined by PEP 518
   build-backend = "flit_core.api" # Defined by PEP 517

> Build-backend Tools

Reviewer: ```suggestion
Build Backend Tools
```

===================

> This section collects some of the most popular build systems available in the

Reviewer: ```suggestion
This section describes some of the most popular build systems in the
```

Python ecosystem. Although all of them achieve the same goal, there are a few
differences regarding their capabilities and the way of specifying project
metadata.

.. TODO: Include links to each build system allowed metadata fields

Setuptools
----------

`Setuptools`_ has been around for a long time in the Python ecosystem. Unless
you require from a high control over the installation steps of your project,
`flit`_ and `poetry`_ should be used.

> If you do not need from a dynamic installation process, you may consider using a
> ``setup.cfg`` file. However, the ``setup.py`` is still required. This file
> should have a call to the ``setup()`` function to act as the entry point of the
> build backend system.

Reviewer: ```suggestion
If you do not need a dynamic installation process, you can consider using a
``setup.cfg`` file. However, the ``setup.py`` file is still required. The ''setup.cfg`` file
should have a call to the ``setup()`` function to act as the entry point of the
build backend system.
```

> All the following `setuptools metadata fields`_ are valid ones. These need to be
> specified either in the ``setup.py`` or ``setup.cfg``.

Reviewer: ```suggestion
All of the following `setuptools metadata fields`_ are valid and must be
specified either in the ``setup.py`` or ``setup.cfg`` file.
```

Flit
----

Flit is a modern and lightweight build system that requires developers
to manage virtual environments on their own. Developers must:

* Create a virtual environment and activate it.
* Install the package in editable mode.

Flit is the default tool for creating a new ``pyansys`` project when using the
`ansys-templates tool`_.

The ``[project]`` section specifies the project's metadata and required
dependencies. For more information, see `flit pyproject.toml
guidelines`_.

Poetry
------

Because of its ``poetry.lock`` file, Poetry provides strong dependency pinning. When
installing a package, poetry creates a virtual environment, thus ensuring an isolated
package development environment.

> Nevertheless, it is possible to make Poetry ignore the ``poetry.lock`` file by running:

Reviewer: ```suggestion
Nevertheless, it is possible to make Poetry ignore the ``poetry.lock`` file with:
```

.. code:: bash

   poetry config virtualenvs.create false --local

Using `poetry`_ is popular because it:

* Supports pinning dependency versions via a ``poetry.lock`` file that can be
  used for testing and CI
* Allows downstream packages to still consume a loose dependency specification
* Integrates with `dependabot`_ to update the pinned version

> The ``[tool.poetry]`` section contains metadata and defines the project's

Reviewer: ```suggestion
The ``[tool.poetry]`` section contains metadata and defines project
```
